File: 2017/09/stream_processing.py
import fileinput


def get_stream_score(s):
    level, score, removed = 0, 0, 0
    skip, garbage = False, False

    for c in s:
        if skip:
            skip = False
            continue

        if c == '!':
            skip = True
            continue

        if c == ">":
            garbage = False
        if garbage:
            removed += 1
            continue
        if c == '<':
            garbage = True

        if c == '{':
            level += 1
        if c == '}':
            score += level
            level -= 1
    
    return score, removed

if __name__ == '__main__':
    for line in fileinput.input():
        print(get_stream_score(line))

File: 2017/09/stream_processing.go
package main

import (
	"bufio"
	"fmt"
	"os"
)

func getStreamScore(s string) (score, removed int) {
	var level int
	skip := false
	garbage := false
	for _, c := range s {
		if skip {
			skip = false
			continue
		}

		if c == '!' {
			skip = true
			continue
		}

		if c == '>' {
			garbage = false
		}
		if garbage {
			removed++
			continue
		}
		if c == '<' {
			garbage = true
		}

		if c == '{' {
			level++
		}
		if c == '}' {
			score += level
			level--
		}
	}
	return
}

func main() {

	scanner := bufio.NewScanner(os.Stdin)

	for scanner.Scan() {
		fmt.Println(getStreamScore(scanner.Text()))
	}

}
